File: solutions/python/knapsack/1/knapsack.py
from itertools import combinations


def maximum_value(maximum_weight, items):
    if maximum_weight <= 0:
        raise ValueError()
    
    if len(items) == 0:
        return 0
    
    best_value = 0
    for i in range(1, len(items)):
        for comb in combinations(items, i):
            weight_comb = sum(item["weight"] for item in comb) 
            if weight_comb > maximum_weight:
                continue
            value_comb = sum(item["value"] for item in comb)
            if value_comb > best_value:
                best_value = value_comb
    return best_value

File: solutions/java/knapsack/1/src/main/java/Knapsack.java
import java.util.List;

class Knapsack {

    private static int maximumValueHelper(int maximumWeight, List<Item> items, int index) {
        int n = items.size();
        if (maximumWeight <= 0) return 0;
        int currentMaxValue = 0;
        for (int newIndex = index; newIndex < n; newIndex++) {
            Item currentItem = items.get(newIndex);
            if (currentItem.weight > maximumWeight) continue;
            int currentValue = currentItem.value 
                + maximumValueHelper(maximumWeight - currentItem.weight, items, newIndex + 1);
            if (currentValue > currentMaxValue) currentMaxValue = currentValue;
        }
        return currentMaxValue;
    }

    public int maximumValue(int maximumWeight, List<Item> items) {
        return maximumValueHelper(maximumWeight, items, 0);
    }

}
